content: Partial update of positioning systems integrations (#82)
fictionlab/docs: docs/integrations/positioning-systems/ublox-evk-m8n.mdx

This tutorial will guide you through the process of connecting a UBlox GPS
module to your Leo Rover.

EVK-7 / EVK-8 / EVK-M8 evaluation kits make evaluating the high performance
u-blox 7, u-blox 8 and u-blox M8 positioning products simple. The built-in USB
interface provides both power supply and high-speed data transfer, and
eliminates the need for an external power supply. The u-blox evaluation kits are
compact, and their user-friendly interface and power supply make them ideally
suited for use in laboratories, vehicles and outdoor locations.

In this tutorial we will present the process of connecting a UBlox GPS module to
your Leo Rover and integrating it with the system on the **UBlox EVK-M8N GPS
module** example.

The steps might slightly differ for other UBlox models, but should be
essentially similar.

## What to expect?

After completing this tutorial, you should be able to gather the GPS data and
get it published on ROS topics.

<ImageZoom
  loading="eager"
  src="/img/robots/leo/integrations/ublox-gps/ech-ublox-gps-node.webp"
  width="294"
  height="431"
  figStyle={{
    width: 400,
  }}
/>

## Prerequisites

> 

Reviewer: Add TODO comment to change the link once ROS2 development tutorial is available

Author: Done

Reviewer: This comment should only be visible in markdown, not on the rendered page

<LinkButton
  to="/leo-rover/1.8/advanced-guides/ros-development"
  title="ROS Development"
  description="Detailed guide on ROS development for Leo Rover, covering topics like adding
  additional functionalities, building ROS packages and more."
/>

## Referenced products

<ProductPreview
  shopUrl="https://www.leorover.tech/shop/gps-module"
  imageSrc="/img/robots/leo/integrations/ublox-gps/ublox-evk-m8n-shop.webp"
  width={1200}
  height={802}
  title="UBlox EVK-M8N GPS module"
  description="Enhance Leo Rover's capabilities with the UBlox EVK-M8N GPS module.
    Experience precise navigation and versatile connectivity with multi-GNSS support,
    ensuring high-performance and reliability."

> />

Reviewer: Move the product preview to `src/products`

Author: Ok

## List of components

- UBlox GPS module (we have used the **UBlox EVK-M8N GPS module**)
- Any computer which you can connect to the rover via ssh.

## Mechanical integration

When it comes to mounting the module, any solution that works for you is okay.
You can 3D print a special mounting plate, use a velcro or even a tape.

## Wiring and electronics connection

Wiring up the module is very simple, as it doesn't require external power
supply. You just need to connect the module with a USB cable with the rover via
the USB socket in the mounting plate. There is a micro-usb mounting hole with
the USB sign on the module, and you need to plug in the USB cable there.

## Software integration

> The first thing you can do is to make sure your device has the correct
> permissions and is available at the fixed path on your system. To do this, you
> can add the following rule to the `udev` service. We will need `idVendor` and
> `idProduct`, to make the rules for the correct device. To get that information,
> you need to connect to the rover via `ssh`, and then, you can use the `lsusb`
> command. It will list all connected usb devices. In the output, you should look
> for something named like **U-Blox AG [u-blox 8]**:

Reviewer: This is quite a long paragraph
```suggestion
The first thing you can do is to make sure your device has the correct
permissions and is available at the fixed path on your system.

We will need USB Vendor and Product IDs to make the rules for the correct device. To get that information,
you need to connect to the rover via `ssh`, and then, you can use the `lsusb`
command. It will list all connected usb devices. In the output, you should look
for something named like **U-Blox AG [u-blox 8]**:
```

Author: Done

<ImageZoom
  src="/img/robots/leo/integrations/ublox-gps/terminal-lsusb.webp"
  width="587"
  height="112"
/>

> Now, our `idVendor` and `idProduct` are the two strings after "ID" divided by

Reviewer: ```suggestion
Now, our Vendor and Product IDs are the two strings after "ID" divided by
```

Author: Done

colon. In our case, `idVendor` is `1546` and `idProduct` is `01a8`.

> After getting those two, the only thing that is left is the port that the device
> connects to. You can get it by running in the terminal the command:
> 
> ```bash
> dmesg -w
> ```
> 
> and then connecting the module to the rover. You should see that the new device
> is registered, and on what port (with this command, you can also get the
> `idVendor` and `idProduct` values).
> 
> <ImageZoom
>   src="/img/robots/leo/integrations/ublox-gps/terminal-ttyacm0.webp"
>   width="867"
>   height="112"
> />

Reviewer: I don't think we need it since our udev rules makes a symlink
```suggestion
```

Author: Okay

Now, you need to make a `udev` rules file.

```xml title="/etc/udev/rules.d/ublox.rules"

> KERNEL=="ttyACM*", ATTRS{idVendor}=="1546", ATTRS{idProduct}=="01a8", MODE="0666", GROUP="dialout", SYMLINK+="ublox_gps"

Reviewer: We can omit the KERNEL check
```suggestion
ATTRS{idVendor}=="1546", ATTRS{idProduct}=="01a8", MODE="0666", GROUP="dialout", SYMLINK+="ublox_gps"
```

Author: Sure.

    <param from="$(find-pkg-share ublox_gps)/config/c94_m8p_rover.yaml" />
    <param name="device" value="/dev/ublox_gps"/>
    <param name="frame_id" value="ublox_frame"/>
  </node>
</launch>
```

In `/etc/ros/robot.launch.xml` add this line somewhere before the closing
`</launch>` tag, so that your node will start at boot.

```xml title="/etc/ros/robot.launch.xml"
<include file="/etc/ros/gps.launch.xml"/>
```

Your robot should be aware of where the module is located and what space it
occupies. You can ensure it does that by creating a URDF model of the module.

```xml title="/etc/ros/urdf/ublox.urdf"
<?xml version="1.0"?>
<robot>
  <link name="ublox_frame"/>

  <joint name="ublox_joint" type="fixed">
    <origin xyz="-0.05 0 0"/>
    <parent link="base_link"/>
    <child link="ublox_frame"/>
  </joint>

</robot>
```

:::warning

Make sure that `frame_id` param in the launch file and link `name` in
`ublox.urdf` file have the same value.

:::

And including it in the description that is uploaded at boot. Add this line
somewhere before the closing `</robot>` tag.

```xml title="/etc/ros/urdf/robot.urdf.xacro"
<xacro:include filename="/etc/ros/urdf/ublox.urdf"/>
```

> The last step is to either reboot the robot
> 
> ```bash
> reboot
> ```
> 
> or restart the nodes
> 
> ```bash
> ros-nodes-restart
> ```

Reviewer: IMO it is not worth recommending reboot.
```suggestion
The last step is to restart the nodes

```bash
ros-nodes-restart
```
```

Author: OKay, but it's mentioned probably in every tutorial.
